### ISSUE TITLE
[Steam Deck] Is it possible to make it autorun while in gaming mode?

### ISSUE THREAD

**Issue by lyndonguitar** [zocker-160/SyncThingy#11]
so no need to go to desktop mode

**Comment by lyndonguitar** [comment]
maybe natively implement this ?
https://www.reddit.com/r/SteamDeck/comments/vocyi5/start_syncthing_automatically_on_steamdeck_even/?sort=new

**Comment by zocker-160** [comment]
Yes I was thinking about this as well, I will definitely take a look.

The downside of this will be, that there won't we an easy way to stop it from game mode and / or check if it is still running.

**Comment by zLeqitEclipse** [comment]
> Yes I was thinking about this as well, I will definitely take a look.
> 
> The downside of this will be, that there won't we an easy way to stop it from game mode and / or check if it is still running.

Well, at least you have the option to fire up chrome in Gaming Mode and access the web interface.

**Comment by zocker-160** [comment]
> Well, at least you have the option to fire up chrome in Gaming Mode and access the web interface.

That is true and probably the best solution. I did think about this problem a fair bit and sadly it looks like that there is no easy / convenient solution to make this work without the user having to run at least one console command.

**Comment by zLeqitEclipse** [comment]
Maybe a function in the system tray called 'Install Service' (or something like that) that runs something like `systemctl --user enable --now syncthingy.service`
Obviously the Program needs to create the service file beforehand.

**Comment by zocker-160** [comment]
That is sadly not possible, the Flatpak does not allow me to run commands on the host. That will always have to be done by the user manually.

There is a way to give execute permissions, but that will not get approved on FlatHub if my only reason to request it is "user convenience". :(

**Comment by xamindar** [comment]
You can create your own systemd service file and launch the flatpak syncthing like so:
(deck@xamdeck user)$ cat /home/deck/.config/systemd/user/syncthing.service

> [Unit]
> Description=Syncthing
> 
> [Service]
> Type=simple
> #ExecStart=flatpak run --command=syncthing me.kozec.syncthingtk
> ExecStart=flatpak run --branch=stable --arch=x86_64 --command=syncthing com.github.zocker_160.SyncThingy
> Restart=on-failure
> RestartSec=1
> SuccessExitStatus=3 4
> 
> [Install]
> WantedBy=default.target
> (deck@xamdeck user)$ 

Then enable and start the service with "systemctl --user enable --now syncthing".

However, it appears the dev has not accounted for syncthing already running. As a result, the SyncThingy tray program (in desktop mode) will open for a few seconds, then close without any indication of why. Best to probably use the syncthing-gtk flatpak for now as that one is designed to manage an already running syncthing instance.

**Comment by jmqm** [comment]
@xamindar , @lyndonguitar wants a convenient way to do it in gaming mode, different from creating a systemd file in desktop mode

**Comment by xamindar** [comment]
> @xamindar , @lyndonguitar wants a convenient way to do it in gaming mode, different from creating a systemd file in desktop mode

@jmqm , then please provide info on where the systray is located in gaming mode or how to autostart background services through steam. Until then, running it as a background service is the only way to make it autorun in gaming mode. Also, maybe try posting something actually useful next time.

**Comment by jmqm** [comment]
@xamindar yikes, you get offended for no reason? There is no convenient way of doing it so original poster is asking if it can be added. What's your problem?

The systemd instruction is in the Reddit post mentioned previously so you didn't provide anything useful, just minor change to the command.

And don't tell me what to do :)

**Comment by mrexodia** [comment]
Maybe you could have a button 'Install service' that just shows a dialog with the steps to follow:

- Open terminal (with link that opens the terminal app? not sure if possible)
- Copy this command -> copies one liner that does the service setup
- Tell the user to run the copied command in the terminal

This should be doable with just the touch screen of the Steam Deck, provided the buttons are big enough.

**Comment by zocker-160** [comment]
here a quick summary of the problem:

1. I am currently not supporting the use case of having syncthing running in the background
2. SyncThingy closes on start silently in that case, because the port is already blocked
3. I cannot register and run a systemd service from inside flatpak (user action will always be required)
4. There is no simple way to manage (aka start / stop / check status etc) Syncthing from game mode as there is no such thing as a tray and getting an entry into the global Steam Deck overlay menu is not trivial.

solutions I can offer:

1.1  I can look into a way to check if SyncThingy flatpak is already running (including as service) and make the systray show up as it should
1.2  I am not interested in supporting the use case of having any syncthing process running somewhere in the system and attach to that as it will cause a lot of problems mainly because:
1.2.1 the "other" syncthing instance will have different settings than the "internal" one shipped with the flatpak
1.2.2 I cannot control the version of the "other" syncthing instance

2. Similar to (1) and I can implement an error message in case the port is blocked by non SyncThingy process
3. I can create the systemd service file when the user requests it and then show a command the user then has to run manually in the terminal
4. I cannot think of any solution for this apart from opening a browser in game mode and navigating to the syncthing webui

thoughts?

**Comment by xamindar** [comment]
I think definitely implementing 1 and possibly 4 would be a good idea. That would cover the "want to run syncthing all the time" crowd.

**Comment by zocker-160** [comment]
Yes I know it has been over 2 month, but I finally found the time to implement this.

I have implemented from the numbers above:

- 1 using the health check API that Syncthing provides
- 3 is kinda implemented, it will show a message with `Syncthing failed to start! exit code (1) `
- 4 is kinda implemented, I added a command to the settings which can be copied and that will use the systemd unit file I am providing in this repository, with this approach I am not misusing the folder permissions granted.
It uses `wget` to pull the file, which IIRC is preinstalled on SteamOS.

I will be pushing this new version to flathub-beta first for further testing.
